firewood/src/storage/linear/proposed.rs

```diff
@@ -0,0 +1,526 @@
+    offset: u64,
+    state: LayeredReaderState<'a>,
+    layer: &'a Proposed<P, M>,
+}
+
+/// A [LayeredReaderState] keeps track of when the next transition
+/// happens for a layer. If you attempt to read bytes past the
+/// transition, you'll get what is left, and the state will change
+#[derive(Debug)]
+enum LayeredReaderState<'a> {
+    // we know nothing, we might already be inside a modified area, one might be coming, or there aren't any left
+    Initial,
+    // we know we are not in a modified area, so find the next modified area
+    FindNext,
+    // we know there are no more modified areas past our current offset
+    NoMoreModifiedAreas,
+    BeforeModifiedArea {
+        next_offset: u64,
+        next_modified_area: &'a [u8],
+    },
+    InsideModifiedArea {
+        area: &'a [u8],
+        offset_within: usize,
+    },
+}
+
+impl<P: ReadLinearStore, M: Send + Sync + Debug> ReadLinearStore for Proposed<P, M> {
+    fn stream_from(&self, addr: u64) -> Result<Box<dyn Read + '_>, Error> {
+        Ok(Box::new(LayeredReader {
+            offset: addr,
+            state: LayeredReaderState::Initial,
+            layer: self,
+        }))
+    }
+
+    fn size(&self) -> Result<u64, Error> {
+        // start with the parent size
+        let parent_size = self.parent.state.size()?;
+        // look at the last delta, if any, and see if it will extend the file
+        Ok(self
+            .new
+            .range(..)
+            .next_back()
+            .map(|(k, v)| *k + v.len() as u64)
+            .map_or_else(
+                || parent_size,
+                |delta_end| std::cmp::max(parent_size, delta_end),
+            ))
+    }
+}
+
+/// Marker that the Proposal is mutable
+#[derive(Debug)]
+pub(crate) struct Mutable;
+
+/// Marker that the Proposal is immutable
+#[derive(Debug)]
+pub(crate) struct Immutable;
+
+impl<P: ReadLinearStore> WriteLinearStore for Proposed<P, Mutable> {
+    // TODO: we might be able to optimize the case where we keep adding data to
+    // the end of the file by not coalescing left. We'd have to change the assumption
+    // in the reader that when you reach the end of a modified region, you're always
+    // in an unmodified region
+
+    fn write(&mut self, offset: u64, object: &[u8]) -> Result<usize, Error> {
+        // the structure of what will eventually be inserted
+        struct InsertData {
+            offset: u64,
+            data: Box<[u8]>,
+        }
+
+        debug_assert!(!object.is_empty());
+        let mut insert = InsertData {
+            offset,
+            data: Box::from(object),
+        };
+
+        // we track areas to delete here, which happens when we combine
+        let mut merged_offsets_to_delete = Vec::new();
+
+        // TODO: combining a change where we constantly append to a file might be too
+        // expensive, so consider skipping the relocations if they don't overlap at all
+        for existing in self.new.range(..=insert.offset).rev() {
```

Comment:
Does this for loop need to iterate more than once? Don't we only need to consider the first diff that this iterator returns? Is it possible any subsequent diff overlaps with the insert?

Reply:
Yeah it should be impossible with the current implementation. I think this code should just disappear though. I'm not sure it's worth coalescing left.

Consider the bulk insert case -- we keep appending new nodes. I don't want to keep reallocating all the stuff we are about to write repeatedly. It's almost like this makes sense to do sometimes. Figuring out what sometimes is seems challenging.

Reply:
Consensus is that we shouldn't coalesce blocks that do not overlap (i.e., adjacent blocks) on either side.